rfd/0073-kube-secret-storage.md

```diff
@@ -0,0 +1,501 @@
+
+### Secret creation and lifecycle
+
+In order to use the secret storage, the secret can be created by [Helm chart](https://github.com/gravitational/teleport/tree/master/examples/chart/teleport-kube-agent) at install time. This is a requirement for the cleanup procedures once the user run `helm uninstall {release}`.
+
+In order to support the secret creation via Helm, the Helm chart has to be changed to reuse the Secret data if it exists in Kubernetes and, if does not exist, create an empty secret. This can be achieved with the Helm's native function `lookup`.
+
+```helm
+{{- $secret := lookup "v1" "Secret" .Release.Namespace "{{ .Release.Name }}-identity" -}}
+{{- if $secret -}}
+{{/*
+   Reusing secret data
+*/}}
+{{- else -}}
+{{/*
+    create secret with empty data
+*/}}
+{{- end -}}
+
+```
+
+#### Secret content
+
+Once Kube Agent updates the secret, it will have the following structure:
+
+```yaml
+apiVersion: v1
+data: |
+    {{$TELEPORT_REPLICA_NAME}}:
+        {{.Values.kubeClusterName}}:
+            {
+            "kind": "identity",
+            "version": "v2",
+            "metadata": {
+                "name": "current"
+            },
+            "spec": {
+                "key": "{key_content}",
+                "ssh_cert": "{ssh_cert_content}",
+                "tls_cert": "{tls_cert_content}",
+                "tls_ca_certs": ["tls_ca_certs"],
+                "ssh_ca_certs": ["ssh_ca_certs"]
+            }
+            }
+// if running in high availabilty mode
+    {{$TELEPORT_REPLICA_NAME[i]}}:
+        {{.Values.kubeClusterName}}:
+            {
+            "kind": "identity",
+            "version": "v2",
+            "metadata": {
+                "name": "current"
+            },
+            "spec": {
+                "key": "{key_content}",
+                "ssh_cert": "{ssh_cert_content}",
+                "tls_cert": "{tls_cert_content}",
+                "tls_ca_certs": ["tls_ca_certs"],
+                "ssh_ca_certs": ["ssh_ca_certs"]
+            }
+            }
+    
+kind: Secret
+metadata:
+  name: {.Release.Name}-identity
+  namespace: {.Release.Namespace}
+```
+
+Where:
+
+- `ssh_cert` is a PEM encoded SSH host cert.
+- `key` is a PEM encoded private key.
+- `tls_cert` is a PEM encoded x509 client certificate.
+- `tls_ca_certs` is a list of PEM encoded x509 certificate of the certificate authority of the cluster.
+- `ssh_ca_certs` is a list of SSH certificate authorities encoded in the authorized_keys format.
+- `TELEPORT_REPLICA_NAME` is the teleport agent replica name. Constant when using Deployments, `TELEPORT_REPLICA_NAME={{ .Release.Name}}` or dynamic when using StatefullSets, `TELEPORT_REPLICA_NAME=metadata.name`.
+
+#### RBAC Changes
+
+The Teleport Kube Agent service account must be able to read and edit secrets within the namespace that is running, therefore, one must create a new namespace role and attach it to the Kube Agent service account with the following content:
```

Comment:
Can we restrict these permissions to only allow the agents read secrets containing identities rather than all secrets?

Reply:
Yes, we can restrict the RBAC with `resourceNames: ["secretName1", "secretName2"]` but the secret must already exist.

`Note: You cannot restrict create or deletecollection requests by their resource name. For create, this limitation is because the name of the new object may not be known at authorization time. If you restrict list or watch by resourceName, clients must include a metadata.name field selector in their list or watch request that matches the specified resourceName in order to be authorized. For example, kubectl get configmaps --field-selector=metadata.name=my-configmap`
Source: https://kubernetes.io/docs/reference/access-authn-authz/rbac/